File: lecture6/cohesion.js
// metrics/cohesion.js
function calculateCohesion(code) {
    const functionRegex = /function\s+(\w+)\s*\([^)]*\)\s*{([\s\S]*?)}/g;
    const variableRegex = /\b(\w+)\b/g;

    const functions = [];
    let match;
    
    // Find all functions
    while ((match = functionRegex.exec(code)) !== null) {
        const functionName = match[1];
        const functionBody = match[2];
        const variables = [];
        
        // Find all variables in the function
        let variableMatch;
        while ((variableMatch = variableRegex.exec(functionBody)) !== null) {
            variables.push(variableMatch[1]);
        }

        functions.push({ functionName, variables });
    }

    // Calculate cohesion based on shared variables
    let sharedVariables = 0;
    for (let i = 0; i < functions.length; i++) {
        for (let j = i + 1; j < functions.length; j++) {
            const shared = functions[i].variables.filter(variable => functions[j].variables.includes(variable)).length;
            sharedVariables += shared;
        }
    }

    return sharedVariables;
}

module.exports = calculateCohesion;

File: lecture6/cohesion.py
# metrics/cohesion.py

import re

def calculate_cohesion(code):
    function_pattern = r'function\s+(\w+)\s*\([^)]*\)\s*{([\s\S]*?)}'
    variable_pattern = r'\b(\w+)\b'

    functions = []
    matches = re.finditer(function_pattern, code)

    for match in matches:
        function_body = match.group(2)
        variables = re.findall(variable_pattern, function_body)
        functions.append(set(variables))

    shared_variables = 0
    for i in range(len(functions)):
        for j in range(i + 1, len(functions)):
            shared = functions[i].intersection(functions[j])
            shared_variables += len(shared)

    return shared_variables
